lib/kdtree.h

```diff
@@ -219,6 +221,36 @@ class FlatNode {
         return static_cast<uint32_t>(data_) >> 2;
     }
 
+    // Since right nodes are create after parent nodes, we need a setter to
+    // update the index of the right node after creation.
+    void set_right(uint32_t value) {
+        data_ = (data_ & 0x0000) |
+                static_cast<uint64_t>(value << 2 |
+                                      static_cast<uint32_t>(split_axis()));
+    }
+
+    // leaf node attributes
+
+    uint32_t first_triangle_id() const {
+        assert(is_leaf());
+        return static_cast<uint32_t>(data_ >> 32);
+    }
+
+    uint32_t second_triangle_id() const {
+        assert(is_leaf());
+        return static_cast<uint32_t>((data_ & 0xFFFFFFFF) >> 2);
+    }
+
+    bool is_sentinel() const {
+        assert(is_leaf());
+        return second_triangle_id() == INVALID_TRIANGLE_ID;
+    }
+
+    bool is_empty() const {
```

Comment:
It was. I've used before an empty node as a sentinel, till I realised that I can just add an inner node without a parent.